gem/backends.py

```diff
@@ -39,23 +39,36 @@ def _update_user_from_claims(user, claims):
     # If the user doesn't have a profile for some reason make one
     if not hasattr(user, 'profile'):
         user.profile = UserProfile(user=user)
+
+        # TODO: we should be using a more specific site here?
         user.profile.site = Site.objects.get(is_default_site=True)
 
     # Ensure the profile is linked to their auth service account using the uuid
     if user.profile.auth_service_uuid is None:
         user.profile.auth_service_uuid = claims.get("sub")
 
+        # If a user already exists with this username
+        # change that user's username
+        username = claims.get("username")
+        if username:
+            for user in User.objects.filter(username=username):
+                user.username = user.profile.site.pk + '_' + username
+                user.save()
+
     # Synchronise a user's profile data
     user.profile.gender = claims.get("gender", "-").lower()[0]
     date_of_birth = claims.get("birthdate", None)
     if date_of_birth:
         date_of_birth = datetime.strptime(date_of_birth, "%Y-%m-%d").date()
     user.profile.date_of_birth = date_of_birth
     if user.profile.alias is None or user.profile.alias == "":
-        user.profile.alias = user.username
-    elif user.profile.alias != user.username:
-        user.profile.alias = user.username
+        user.profile.alias = user.profile.migrated_username
+    elif user.profile.alias != user.profile.migrated_username:
+        user.profile.alias = user.profile.migrated_username
     user.profile.save()
+    if username:
+        user.username = username
+        user.save()
```

Comment:
I think this block should be up at the top where we change the clashing username.
That way it should only happen when users are reclaiming accounts and not on every login.